File: src/www/__test__/testSteps/board/HasInHisHandNoCards.ts
import { AbstractPostLineStep } from "../AbstractPostLineStep";
import { queryAllByCard, getByHand } from "./helpers";

export class HasInHisHandNoCards extends AbstractPostLineStep {
  constructor() {
    super(/_([^_]+)_ has( in h[ei][rs] hand)? no _([^_]+)_ card(?!s? of)/i);
  }

  async runMatch(_: any, match: string[]) {
    const [, player, , type] = match;
    const boardSquare = getByHand(document.body, player);
    const cards = queryAllByCard(boardSquare, type);
    expect(cards).toHaveLength(0);
  }
}

File: src/test/java/com/drpicox/game/testSteps/board/HasInHisHandNoCards.java
package com.drpicox.game.testSteps.board;

import com.drpicox.game.forms.VisibleGameForm;
import com.drpicox.game.testPost.TestPostForms;
import com.drpicox.game.testPost.reader.PostLine;
import com.drpicox.game.testSteps.AbstractPostLineStep;
import org.springframework.stereotype.Component;

import static com.google.common.truth.Truth.assertThat;

@Component
public class HasInHisHandNoCards extends AbstractPostLineStep {

    private final TestPostForms testPostForms;

    public HasInHisHandNoCards(TestPostForms testPostForms) {
        this.testPostForms = testPostForms;
    }

    @Override
    protected String getRegex() {
        return "_([^_]+)_ has in h[ei][rs] hand no _([^_]+)_ card(?!s? of)";
    }

    @Override
    protected void run(PostLine line, String[] match) {
        var player = match[1];
        var square = 0;
        var expectedCount = 0;
        var type = match[2];

        var response = testPostForms.getForm(VisibleGameForm.class);
        var cards = response.getCards().ofOwner(player).atSquare(square).ofType(type);

        assertThat(cards).isEmpty();
    }
}
